### ISSUE TITLE
Sometimes hangs on watching large trees on FreeBSD

### ISSUE THREAD

**Issue by t6** [idea4bsd/fsnotifier#1]
To reproduce `cd /usr/ports; fsnotifier --selftest`

**Comment by t6** [comment]
`sbwait`:
```
The sbwait wchan is present when a thread has invoked the in-kernel
sbwait() function to wait for a socket event.  It's used in a number of
situations, but the main ones are:

- The thread is trying to send on a blocking socket, but there's
  insufficient socket buffer space, so it must wait for space.  This might
  occur if it has managead to max out the bandwidth available to a TCP
  connection, or flow control is in use and the receiver does not wish to
  receive more data yet.

- The thread is trying to receive on a blocking socket, but there's not
  enough data to satisfy the read request, so it must wait for data to be
  received.  It might be waiting for a remote TCP sender to have data
  available, or for in-flight data to arrive.
```
from: https://lists.freebsd.org/pipermail/freebsd-questions/2005-February/075647.html

**Comment by jirib** [comment]
I see

```
...
fsnotifier[32627]  warn: inotify_add_watch(/home/jirib/cvs/openbsd-ports/archivers/libarchive/CVS): Too many open files
...
```

**Comment by t6** [comment]
Ok, I guess you hit the maxfiles limit in OpenBSD. It's probably set to something smaller than in FreeBSD. Is there a way to increase maxfiles? On FreeBSD I can use `sysctl kern.maxfiles=` to set it. Currently it's set to 782342.

**Comment by t6** [comment]
Fixed it by patching libinotify (see ref. pull request). We'll see if this is ok.

**Comment by jirib** [comment]
Does creating virtualenv (in python plugin) work for you? It does not for me. IMO intellij code has a lot of if win elif linux elif freebsd and that's all :)

**Comment by t6** [comment]
Yes, that works for me. A fast grep of the IntelliJ code reveals several such checks: https://gist.github.com/t6/282bf1f627fdc28bf190

Do you think one of them is the reason it doesn't work on OpenBSD?

**Comment by jirib** [comment]
I'll check it. IMO it is related to spawning processes.

**Comment by jirib** [comment]
Just a question, are you able to build idea from sources?

**Comment by t6** [comment]
Nope. It fails with
```
Error:Kotlin: [Internal Error] java.rmi.ServerError: Error occurred in server thread; nested exception is: 
	java.lang.UnsatisfiedLinkError: Could not load library. Reasons: [no jansi64-1.0.0 in java.library.path, no jansi-1.0.0 in java.library.path, no jansi in java.library.path]
```
There's no jansi support for FreeBSD. Unfortunately the Kotlin compiler relies on it and the IDEA build system just downloads a pre-built Kotlin so patching it directly is tricky. 

Forking Kotlin and removing all jansi references should be possible (and then patching IDEA's build system to download the patched version instead). There doesn't seem to be many jansi references in Kotlin's source tree.

**Comment by t6** [comment]
I've created https://github.com/idea4bsd/kotlin-patch which patches `kotlin-compiler.jar` to not depend on jansi (copy the result `target/kotlin-compiler.jar` over `build/kotlinc/lib/kotlin-compiler.jar` in IDEA's source tree if you build with ant or the bundled Kotlin in `/usr/local/intellij/plugins/Kotlin/kotlinc/lib/kotlin-compiler.jar`).

The build goes further now, but fails because it cannot find javafx. I still need to figure out how to stop the build of trying to build everything in the repo.

On the other hand patching IDEA's jars directly like I did with Kotlin might not be as horrible as I initially thought.

**Comment by t6** [comment]
FYI with JavaFX I'm able to build my own IDEA releases.

**Comment by mnd999** [comment]
I still see regular hangs in IntelliJ using FreeBSD with this notifier. Doing killall fsnotifier frees up IntelliJ and it seems to respawn.

**Comment by t6** [comment]
Are you using the port `java/intellij-fsnotifier` or do you compile it yourself? If compiling it yourself do you use the libinotify fixes from @wulf7 (see linked issue)? The ports tree does not include the fixes atm, since they haven't been merged yet.

There is an fsnotifier version here that is statically linked against libinotify with @wulf7's fixes applied which you could try: https://github.com/idea4bsd/idea4bsd/raw/idea4bsd-master/bin/freebsd/fsnotifier64

Using that version makes the hangs go way for me, but `fsnotifier` still dies sometimes. At least IntelliJ will then respawn it without needing to use `killall fsnotifer`.

**Comment by wulf7** [comment]
> but fsnotifier still dies sometimes

I quickly glanced over fsnotifier code and see at least one issue which can trigger SIGSEGV. fsnotifier misses support for partial inotify event reads from inotifyfd. It cannot happen with Linux but can  happen with libinotify
To workaround it you can try to bump EVENT_BUF_LEN value to value of net.local.stream.sendspace sysctl (8192 on my laptop) and use latest libinotify (port version is ok).
EVENT_BUF_LEN is defined at line 60 of inotify.c
